examples/step-78/step-78.cc

```diff
@@ -0,0 +1,898 @@
+            forcing_terms *= time_step * (1 - theta);
+            system_rhs -= forcing_terms;
+
+            // Next, we add the forcing terms to the ones that come from the
+            // time stepping, and also build the matrix $\left[\bold{M}+
+            // \frac{1}{4}k_n\sigma^2\bold{D}+k_nr\bold{M}\right]$ that we have
+            // to invert in each time step. The final piece of these operations
+            // is to eliminate hanging node constrained degrees of freedom from
+            // the linear system:
+            system_matrix.copy_from(mass_matrix);
+            system_matrix.add(
+              (theta)*time_step *
+                Utilities::fixed_power<2, double>(asset_volatility) * 0.5,
+              laplace_matrix);
+            system_matrix.add((time_step)*interest_rate * theta * (1 + 1),
+                              mass_matrix);
+
+            constraints.condense(system_matrix, system_rhs);
+
+            // There is one more operation we need to do before we can solve it:
+            // boundary values. To this end, we create a boundary value object,
+            // set the proper time to the one of the current time step, and
+            // evaluate it as we have done many times before. The result is used
+            //  to also set the correct boundary values in the linear system:
+            {
+              RightBoundaryValues<dim> right_boundary_function(strike_price,
+                                                               interest_rate);
+              LeftBoundaryValues<dim>  left_boundary_function;
+              right_boundary_function.set_time(time);
+              left_boundary_function.set_time(time);
+              std::map<types::global_dof_index, double> boundary_values;
+              VectorTools::interpolate_boundary_values(dof_handler,
+                                                       0,
+                                                       left_boundary_function,
+                                                       boundary_values);
+              VectorTools::interpolate_boundary_values(dof_handler,
+                                                       1,
+                                                       right_boundary_function,
+                                                       boundary_values);
+              MatrixTools::apply_boundary_values(boundary_values,
+                                                 system_matrix,
+                                                 solution,
+                                                 system_rhs);
+            }
+
+            // With this out of the way, all we have to do is solve the system,
+            // generate graphical data on the last cycle, and create the
+            // convergence table data.
+            solve_time_step();
+
+            if (cycle == (n_cycles - 1))
+              {
+                add_results_for_output();
+              }
```

Comment:
On lines 726-729, I do this because I need to save the initial conditions in the data_out_stack. Then in lines 837-840, I save each time-step. So I think I needed to do it twice. How did you think this should be done?

Reply:
Oh for some reason I thought they were both part of the loop. This makes sense now.